File: frontend/src/features/users/services/userService.ts
import { apiClient } from "../../../lib/axios";
import type {
    CreateUserRequest,
    UpdateUserContactRequest,
    UpdateUserStatusRequest,
    UserDetail,
    UserListItem,
} from "../types/user.types";

export async function getUsers(): Promise<UserListItem[]> {
    const response = await apiClient.get<UserListItem[]>("/api/users");
    return response.data;
}

export async function getUserById(id: number): Promise<UserDetail> {
    const response = await apiClient.get<UserDetail>(`/api/users/${id}`);
    return response.data;
}

export async function getMyProfile(): Promise<UserDetail> {
    const response = await apiClient.get<UserDetail>("/api/users/me");
    return response.data;
}

export async function createUser(
    request: CreateUserRequest
): Promise<UserDetail> {
    const response = await apiClient.post<UserDetail>("/api/users", request);
    return response.data;
}

export async function updateUserStatus(
    id: number,
    request: UpdateUserStatusRequest
): Promise<UserDetail> {
    const response = await apiClient.patch<UserDetail>(
        `/api/users/${id}/status`,
        request
    );

    return response.data;
}

export async function updateUserContact(
    id: number,
    request: UpdateUserContactRequest
): Promise<UserDetail> {
    const response = await apiClient.patch<UserDetail>(
        `/api/users/${id}/contact`,
        request
    );

    return response.data;
}

File: backend/user-service/src/main/java/com/arizanet/userservice/service/UserService.java
package com.arizanet.userservice.service;

import com.arizanet.userservice.dto.request.CreateUserRequest;
import com.arizanet.userservice.dto.request.UpdateUserContactRequest;
import com.arizanet.userservice.dto.request.UpdateUserStatusRequest;
import com.arizanet.userservice.dto.response.UserDetailResponse;
import com.arizanet.userservice.dto.response.UserListResponse;
import com.arizanet.userservice.entity.User;
import com.arizanet.userservice.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;

    public List<UserListResponse> getAllUsers() {
        return userRepository.findAll()
                .stream()
                .map(this::mapToListResponse)
                .toList();
    }

    public UserDetailResponse getUserById(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Kullanıcı bulunamadı. ID: " + id));

        return mapToDetailResponse(user);
    }

    public UserDetailResponse getMyProfile(String username) {
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("Kullanıcı bulunamadı. Kullanıcı adı: " + username));

        return mapToDetailResponse(user);
    }

    public UserDetailResponse createUser(CreateUserRequest request) {
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new RuntimeException("Bu kullanıcı adı zaten kullanılıyor: " + request.getUsername());
        }

        User user = new User();
        user.setFullName(request.getFullName());
        user.setUsername(request.getUsername());
        user.setPasswordHash(request.getPassword());
        user.setEmail(request.getEmail());
        user.setPhone(request.getPhone());
        user.setRole(request.getRole());
        user.setActive(true);
        user.setStatus(request.getStatus());
        user.setCreatedAt(LocalDateTime.now());

        User savedUser = userRepository.save(user);

        return mapToDetailResponse(savedUser);
    }

    public UserDetailResponse updateUserStatus(Long id, UpdateUserStatusRequest request) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Kullanıcı bulunamadı. ID: " + id));

        user.setActive(request.getActive());

        if (Boolean.TRUE.equals(request.getActive())) {
            user.setStatus("Aktif");
        } else {
            user.setStatus("Pasif");
        }

        User updatedUser = userRepository.save(user);

        return mapToDetailResponse(updatedUser);
    }

    public UserDetailResponse updateUserContact(Long id, UpdateUserContactRequest request) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Kullanici bulunamadi. ID: " + id));

        user.setEmail(request.getEmail());
        user.setPhone(request.getPhone());

        User updatedUser = userRepository.save(user);

        return mapToDetailResponse(updatedUser);
    }

    private UserListResponse mapToListResponse(User user) {
        return new UserListResponse(
                user.getId(),
                user.getFullName(),
                user.getUsername(),
                user.getEmail(),
                user.getPhone(),
                user.getRole(),
                user.isActive(),
                user.getStatus()
        );
    }

    private UserDetailResponse mapToDetailResponse(User user) {
        return new UserDetailResponse(
                user.getId(),
                user.getFullName(),
                user.getUsername(),
                user.getEmail(),
                user.getPhone(),
                user.getRole(),
                user.isActive(),
                user.getStatus(),
                user.getCreatedAt()
        );
    }
}
